microceph/ceph/services.go

```diff
@@ -4,13 +4,128 @@ import (
 	"context"
 	"database/sql"
 	"fmt"
+	"time"
 
+	"github.com/Rican7/retry"
+	"github.com/Rican7/retry/backoff"
+	"github.com/Rican7/retry/strategy"
 	"github.com/canonical/microcluster/state"
+	"github.com/lxc/lxd/shared/logger"
 
 	"github.com/canonical/microceph/microceph/api/types"
 	"github.com/canonical/microceph/microceph/database"
+	"github.com/tidwall/gjson"
 )
 
+type Set map[string]struct{}
+
+func (sub Set) isIn(super Set) bool {
+	flag := true
+
+	// mark flag false if any key from subset is not present in superset.
+	for key := range sub {
+		_, ok := super[key]
+		if !ok {
+			flag = false
+			break // Break the loop.
+		}
+	}
+
+	return flag
+}
+
+// Table to map fetchFunc for workers (daemons) to a service.
+var serviceWorkerTable = map[string](func () (Set, error)) {
+	"osd": getUpOsds,
+	"mon": getMons,
+}
+
+// Restarts (in order) all Ceph Services provided in the input slice on the host.
+func RestartCephServices(services []string) error {
+	for i := range services {
+		err := RestartCephService(services[i])
+		if err != nil {
+			logger.Error(fmt.Sprintf("Service %s restart failed: %v ", services[i], err))
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Restart provided ceph service ("mon"/"osd"...) on the host.
+func RestartCephService(service string) error {
+	if _, ok := serviceWorkerTable[service]; !ok {
+		errStr := fmt.Sprintf("No handler defined for service %s", service)
+		logger.Error(errStr)
+		return fmt.Errorf(errStr)
+	}
+
+	// Fetch a Set{} of available daemons for the service.
+	workers, err := serviceWorkerTable[service]()
+	if err != nil {
+		logger.Errorf("Failed fetching service %s workers", service)
+		return err
+	}
+
+	// Restart the service.
+	snapRestart(service, false)
+
+	// Check all the daemons available before Restart are up.
+	err = retry.Retry(func(i uint) error {
+		iWorkers, err := serviceWorkerTable[service]()
+		if err != nil {
+			return err
```

Comment:
Might add a `logger.Debug()` output logging the outcome of the attempt

Reply:
As discussed, a better place for logging the intermittent failures would be inside the individual handles for ceph services.

Reply:
Done.